openshift-kube-apiserver/filters/deprecatedapirequest/controller.go

```diff
@@ -0,0 +1,239 @@
+		now := time.Now()
+		// clear old entries from processing
+		c.processing.trim(now)
+		// move incoming to processing
+		c.moveIncomingToProcessing()
+		// process processing queue
+		var wg sync.WaitGroup
+		for r := range c.processing.Counts {
+			go func(r string) {
+				wg.Add(1)
+				defer wg.Done()
+				c.processing.Lock()
+				counts := c.processing.Counts[r]
+				c.processing.Unlock()
+				err := c.ensureLogExists(ctx, r)
+				if err != nil {
+					runtime.HandleError(err)
+					return
+				}
+				_, _, err = v1helpers.UpdateStatus(ctx, c.client, r,
+					v1helpers.RotateNodeRequestLogs(c.nodeName, now),
+					IncrementRequestCounts(c.nodeName, counts, c.processing.LastUpdate, now))
+				if err != nil {
+					runtime.HandleError(err)
+					return
+				}
+				c.processing.Lock()
+				c.processing.Counts[r] = map[int]map[string]map[string]int{}
+				c.processing.Unlock()
+			}(r)
+		}
+		wg.Wait()
+	}, c.updatePeriod, stop)
+}
+
+func (c *controller) moveIncomingToProcessing() {
+	if len(c.incoming.Counts) == 0 {
+		return
+	}
+	c.incoming.Lock()
+	defer c.incoming.Unlock()
+	c.processing.IncrementRequestCounts(c.incoming)
+	c.incoming.Counts = map[string]map[int]map[string]map[string]int{}
+}
+
+func (c *controller) ensureLogExists(ctx context.Context, r string) error {
+	_, err := c.client.Get(ctx, r, metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+	if !errors.IsNotFound(err) {
+		return err
+	}
+	log := &apiv1.DeprecatedAPIRequest{
+		ObjectMeta: metav1.ObjectMeta{Name: r},
+		Spec:       apiv1.DeprecatedAPIRequestSpec{RemovedRelease: c.removedRelease(r)},
+	}
+	_, err = c.client.Create(ctx, log, metav1.CreateOptions{})
+	return err
+}
+
+type requestCounts struct {
+	sync.Mutex
+	LastUpdate time.Time
+	Counts     map[string]map[int]map[string]map[string]int
+}
+
+func (rc *requestCounts) IncrementRequestCount(r, verb, user string, ts time.Time, count int) {
+	rc.incrementRequestCount(r, verb, user, ts.Hour(), count)
+	if ts.After(rc.LastUpdate) {
+		rc.LastUpdate = ts
+	}
+}
+
+func (rc *requestCounts) incrementRequestCount(r, verb, user string, h, count int) {
+	if rc.Counts == nil {
+		rc.Counts = map[string]map[int]map[string]map[string]int{}
+	}
+	if _, ok := rc.Counts[r]; !ok {
+		rc.Counts[r] = map[int]map[string]map[string]int{}
+	}
+	if _, ok := rc.Counts[r][h]; !ok {
+		rc.Counts[r][h] = map[string]map[string]int{}
+	}
+	if _, ok := rc.Counts[r][h][user]; !ok {
+		rc.Counts[r][h][user] = map[string]int{}
+	}
+	if _, ok := rc.Counts[r][h][user][verb]; !ok {
+		rc.Counts[r][h][user][verb] = 0
+	}
+	rc.Counts[r][h][user][verb] += count
+}
+
+func (rc *requestCounts) IncrementRequestCounts(o *requestCounts) {
+	for r := range o.Counts {
+		for h := range o.Counts[r] {
+			for u := range o.Counts[r][h] {
+				for v := range o.Counts[r][h][u] {
+					rc.incrementRequestCount(r, v, u, h, o.Counts[r][h][u][v])
+				}
+			}
+		}
+	}
+	if o.LastUpdate.After(rc.LastUpdate) {
+		rc.LastUpdate = o.LastUpdate
```

Comment:
this update moment seems a little odd. Why do we need to track this?  It seems like we would want to track when we flushed to the API, but I'm not sure why we care about the last update time here.  You've already categorized into hours at this point.